apps/prairielearn/src/sync/fromDisk/entity-list.ts

```diff
@@ -0,0 +1,113 @@
+interface NamedEntity {
+  name: string;
+}
+
+type ExistingEntity<Entity> = Entity & {
+  // TODO: make non-nullable once we make this non-null in the database schemas.
+  number: number | null;
+  implicit: boolean;
+};
+
+type ExtraEntity<Entity> = Entity & {
+  implicit: boolean;
+};
+
+type DesiredEntity<Entity> = Entity & { number: number; implicit: boolean };
+
+export function determineOperationsForEntities<Entity extends NamedEntity>({
+  courseEntities,
+  extraEntities,
+  existingEntities,
+  knownNames,
+  makeImplicitEntity,
+  isInfoCourseValid,
+  deleteUnused,
+}: {
+  courseEntities: Entity[];
+  extraEntities?: ExtraEntity<Entity>[];
+  existingEntities: ExistingEntity<Entity>[];
+  knownNames: Set<string>;
+  makeImplicitEntity: (name: string) => Entity;
+  isInfoCourseValid: boolean;
+  deleteUnused: boolean;
+}): {
+  entitiesToCreate: DesiredEntity<Entity>[];
+  entitiesToUpdate: DesiredEntity<Entity>[];
+  entitiesToDelete: string[];
+} {
+  const existingEntityNames = new Set(existingEntities.map((entity) => entity.name));
+  const desiredEntities = new Map<string, DesiredEntity<Entity>>();
+
+  // If `infoCourse.json` is invalid, keep all existing assessment sets in place.
+  // Otherwise, sync whatever is in the JSON file.
+  if (isInfoCourseValid) {
+    for (const entity of courseEntities) {
+      desiredEntities.set(entity.name, {
+        ...entity,
+        implicit: false,
+        number: desiredEntities.size + 1,
+      });
+    }
+  } else {
+    for (const entity of existingEntities) {
+      desiredEntities.set(entity.name, {
+        ...entity,
+        number: desiredEntities.size + 1,
+      });
+    }
+  }
+
+  // Consider each entity name that's actually used. If it doesn't already exist,
+  // add an implicit version. Sort for consistent ordering.
+  for (const name of Array.from(knownNames).sort()) {
+    if (desiredEntities.has(name)) continue;
+
+    desiredEntities.set(name, {
+      ...makeImplicitEntity(name),
+      implicit: true,
+      number: desiredEntities.size + 1,
+    });
+  }
+
+  // Add any extra entities at the end.
+  if (extraEntities?.length) {
+    for (const entity of extraEntities) {
+      // Give precedence to user-provided entities if they have the same name.
+      if (desiredEntities.has(entity.name)) continue;
+
+      desiredEntities.set(entity.name, {
+        ...entity,
+        number: desiredEntities.size + 1,
+      });
+    }
+  }
+
+  // Based on the set of desired assessment entities, determine which ones must be
+  // added, updated, or deleted.
+  const entitiesToAdd = new Map<string, DesiredEntity<Entity>>();
+  const entitiesToUpdate = new Map<string, DesiredEntity<Entity>>();
+  const entitiesToDelete = new Set<string>();
+
+  for (const [name, entity] of desiredEntities) {
+    if (existingEntityNames.has(name)) {
+      // TODO: check for equality, skip update if not needed.
```

Comment:
This would be very nice! But I do agree that the new code is no worse than the old (ON CONFLICT DO UPDATE) code.

Reply:
I did end up implementing this in 55c1b83 (with tests!).

Reply:
Ugh, seems like things are still broken, even with tests. Investigating.

Reply:
Turns out the test itself was broken! Fixed in 1e8ee65.